pkg/scheduler/queue/tenant_queues.go

```diff
@@ -103,28 +156,66 @@ func (qb *queueBroker) makeQueuePath(request *tenantRequest) (QueuePath, error)
 	return QueuePath{string(request.tenantID)}, nil
 }
 
-func (qb *queueBroker) dequeueRequestForQuerier(lastTenantIndex int, querierID QuerierID) (*tenantRequest, *queueTenant, int, error) {
-	tenant, tenantIndex, err := qb.tenantQuerierAssignments.getNextTenantForQuerier(lastTenantIndex, querierID)
-	if tenant == nil || err != nil {
-		return nil, tenant, tenantIndex, err
+func (qb *queueBroker) dequeueRequestForQuerier(
+	lastTenantIndex int,
+	querierID QuerierID,
+) (
+	*tenantRequest,
+	*queueTenant,
+	int,
+	error,
+) {
+	// check if querier is registered and is not shutting down
+	if q := qb.tenantQuerierAssignments.queriersByID[querierID]; q == nil || q.shuttingDown {
+		return nil, nil, qb.tenantQuerierAssignments.tenantOrderIndex, ErrQuerierShuttingDown
 	}
 
-	queuePath := QueuePath{string(tenant.tenantID)}
-	queueElement := qb.tenantQueuesTree.DequeueByPath(queuePath)
-
-	queueNodeAfterDequeue := qb.tenantQueuesTree.getNode(queuePath)
-	if queueNodeAfterDequeue == nil {
-		// queue node was deleted due to being empty after dequeue
-		qb.tenantQuerierAssignments.removeTenant(tenant.tenantID)
+	var queuePath QueuePath
+	var queueElement any
+	if tq, ok := qb.tree.(*TreeQueue); ok {
+		tenant, tenantIndex, err := qb.tenantQuerierAssignments.getNextTenantForQuerier(lastTenantIndex, querierID)
+		if tenant == nil || err != nil {
+			return nil, tenant, tenantIndex, err
+		}
+		qb.tenantQuerierAssignments.tenantOrderIndex = tenantIndex
+		queuePath = QueuePath{string(tenant.tenantID)}
+		queueElement = tq.DequeueByPath(queuePath)
+	} else if itq, ok := qb.tree.(*MultiQueuingAlgorithmTreeQueue); ok {
+		qb.tenantQuerierAssignments.updateQueuingAlgorithmState(querierID, lastTenantIndex)
+		queuePath, queueElement = itq.Dequeue()
 	}
 
 	var request *tenantRequest
+	var tenantID TenantID
 	if queueElement != nil {
 		// re-casting to same type it was enqueued as; panic would indicate a bug
 		request = queueElement.(*tenantRequest)
+		tenantID = request.tenantID
 	}
 
-	return request, tenant, tenantIndex, nil
+	var tenant *queueTenant
+	if tenantID != "" {
+		tenant = qb.tenantQuerierAssignments.tenantsByID[tenantID]
+	}
+
+	// TODO (casie): When deprecating TreeQueue, clean this up.
+	// This cannot be handled by the Tree interface without defining some other, more expansive interfaces
+	// between the legacy and integrated tree queues, which would be more overhead than it's worth, given
+	// that we will eventually retire the legacy tree queue.
+	if tq, ok := qb.tree.(*TreeQueue); ok {
+		queueNodeAfterDequeue := tq.getNode(queuePath)
+		if queueNodeAfterDequeue == nil {
+			// queue node was deleted due to being empty after dequeue
+			qb.tenantQuerierAssignments.removeTenant(tenant.tenantID)
```

Comment:
Yeah, initializing `tenant` on L194 ensures that in the worst case, it's a zero-value (not nil). If `tenantID` isn't in `tenantsByID`, `tenant` will remain assigned to a zero-value.

Reply:
Thinking through this aloud:

* The zero value of a `*queueTenant` is nil, and `tenant.tenantID` will panic if `tenant` is nil.
* `tenant` is nil when `tenantID == ""`
* `tenantID` is `""` when `queueElement` is nil
* `queueElement` is nil when the `tree` is neither a `TreeQueue` nor a `MultiQueingAlgorithmTreeQueue`; OR when `Dequeue()` or `DequeueByPath()` return nil.
* `DequeueByPath()` returns nil here: https://github.com/grafana/mimir/blob/794c3da0a4176916acfc67034a0856ef93d831fe/pkg/scheduler/queue/tree_queue.go#L196
* So it at least seems possible for this to panic, unless I'm missing some guarantee elsewhere

Reply:
> The zero value of a *queueTenant is nil, and tenant.tenantID will panic if tenant is nil.

Ah, you're right; this is a pointer, not a struct, so its zero value is nil. That's my mistake.

> OR when Dequeue() or DequeueByPath() return nil.

**For TreeQueue:** 
This logic relies on the fact that `getNextTenantForQuerier` returns a tenant that is assigned via `tenantsByID`: https://github.com/grafana/mimir/blob/8eafd44449064ea5c140a683081d68b28932fb02/pkg/scheduler/queue/tenant_querier_assignment.go#L117

If _that_ returned `tenant` is nil, we return early. If not, `tenantsByID` is never modified during dequeue (because the TreeQueue doesn't know about it, and, IIUC, removal from `tenantsByID` will never happen in a parallel operation, because we [dispatch operations](https://github.com/grafana/mimir/blob/7d640d9abec2796abf296fa97dffb001e52fff8f/pkg/scheduler/queue/queue.go#L251) in series, so there will never be a concurrent dequeue operation happening).

**For MultiQueuingAlgorithmTreeQueue:**
If objects have been queued correctly (under the tenantID in the request) by the `queueBroker`, the tenant should also be in `tenantsByID`. I think there's a `nil` possibility here, which doesn't matter _within_ `dequeueRequestForQuerier`, but could also cause problems one level up, e.g., in `trySendNextRequestForQuerier`: https://github.com/grafana/mimir/blob/cd2163247af6855e066a9426623c336e2bbf8d2b/pkg/scheduler/queue/queue.go#L397

Again, this is a non-issue, because `tenant` is nil when we were unable to dequeue a `queueElement`, so `req` will also be nil and is caught by this nil check: https://github.com/grafana/mimir/blob/cd2163247af6855e066a9426623c336e2bbf8d2b/pkg/scheduler/queue/queue.go#L352

---
I think this is annoying/difficult to confirm, and doesn't inspire confidence, and we could avoid the confusion by returning early if `queueElement` is nil.

Reply:
> we could avoid the confusion by returning early if queueElement is nil.

I've done this; I hope that's easier to understand -- let me know if there's a different change you think would help.

Reply:
Thanks for such a thorough reply! Makes sense.